lib/spack/spack/test/llnl/util/tty/log.py

```diff
@@ -0,0 +1,472 @@
+    _mock_shell_integration_no_termios(proc, ctl, attrs)
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+def mock_shell_integration_loop_no_termios(proc, ctl, attrs):
+    for i in range(10):
+        _mock_shell_integration_no_termios(proc, ctl, attrs)
+    os.kill(proc.pid, signal.SIGUSR1)
+
+
+@contextlib.contextmanager
+def no_termios():
+    saved = llnl.util.tty.log.termios
+    llnl.util.tty.log.termios = None
+    yield
+    llnl.util.tty.log.termios = saved
+
+
+@contextlib.contextmanager
+def nullcontext():
+    yield
+
+
+@pytest.mark.skipif(not which("ps"), reason="requires ps utility")
+@pytest.mark.skipif(not termios, reason="requires termios support")
+@pytest.mark.parametrize('test_fn,termios_on_or_off', [
+    # tests with termios
+    (mock_shell_fg, nullcontext),
+    (mock_shell_bg, nullcontext),
+    (mock_shell_bg_fg, nullcontext),
+    (mock_shell_fg_bg, nullcontext),
+    (mock_shell_tstp_cont, nullcontext),
+    (mock_shell_tstp_tstp_cont, nullcontext),
+    (mock_shell_tstp_tstp_cont_cont, nullcontext),
+    # tests without termios
+    (mock_shell_fg_no_termios, no_termios),
+    (mock_shell_bg, no_termios),
+    (mock_shell_bg_fg_no_termios, no_termios),
+    (mock_shell_fg_bg_no_termios, no_termios),
+    (mock_shell_tstp_cont, no_termios),
+    (mock_shell_tstp_tstp_cont, no_termios),
+    (mock_shell_tstp_tstp_cont_cont, no_termios),
+])
+def test_foreground_background_simple(test_fn, termios_on_or_off):
+    shell = PseudoShell(test_fn, mock_logger)
+    shell.attrs["debug"] = True
+
+    # run the shell test
+    with termios_on_or_off():
+        shell.start()
+    exitcode = shell.join()
+
+    # processes completed successfully
+    assert exitcode == 0
+
+
+@pytest.mark.skipif(not which("ps"), reason="requires ps utility")
+@pytest.mark.skipif(not termios, reason="requires termios support")
+@pytest.mark.parametrize('test_fn,termios_on_or_off', [
+    # tests with termios
+    (mock_shell_v_v, nullcontext),
+    (mock_shell_integration_1, nullcontext),
+    (mock_shell_integration_loop, nullcontext),
+    # tests without termios
+    (mock_shell_v_v_no_termios, no_termios),
+    (mock_shell_integration_1_no_termios, no_termios),
+    (mock_shell_integration_loop_no_termios, no_termios),
+])
+def test_foreground_background_with_output(test_fn, capfd, termios_on_or_off):
+    shell = PseudoShell(test_fn, mock_logger)
+    shell.attrs["debug"] = True
+
+    with termios_on_or_off():
+        shell.start()
+    exitcode = shell.join()
+
+    out, err = capfd.readouterr()
+    print("stderr ===========================")
+    print(err)
+    print()
+    print("stdout ===========================")
+    print(out)
+    print()
+
+    # processes completed successfully
+    assert exitcode == 0
+
+    # numbers output by logger
+    stripped = out.strip()
+    split = re.split(r'\s+', stripped) if stripped else []
+
+    numbers = set([int(n) for n in split])
+    all_numbers = set(range(shell.attrs["count"]))
```

Comment:
`expected_numbers`? IMO "expected" is common in variable names that are used to store the state we think ought to be.

Reply:
fixed